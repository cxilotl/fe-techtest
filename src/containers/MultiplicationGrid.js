import React, { useState } from 'react';
import PropTypes from "prop-types";
import NumbersGrid from '../components/numbersGrid/NumbersGrid';

const getMultiplesList = (multiplier, maxNum) => {
  let multiple = 1;
  const multiplesList = [];
  if (multiplier > 0 && maxNum >= multiplier) {
    while (multiple < maxNum + 1) {
      const rest = multiple % multiplier;
      if (rest === 0) {
        multiplesList.push(multiple);
      }
      multiple++;
    }
  }
  return multiplesList
};

const MultiplicationGrid = ({ highestMultiplierNumber }) => {
  const [ multiplier, setMultiplier ] = useState(0);
  const multiples = getMultiplesList(multiplier, highestMultiplierNumber);

  const onNumberSelectedFn = (selectedValue) => {
    if (selectedValue === multiplier) {
      setMultiplier(0);
    } else {
      setMultiplier(selectedValue);
    }
  };

  return (
    <>
      <NumbersGrid
        maxNumToDisplay={ highestMultiplierNumber }
        selectedNum={ multiplier }
        highlightedNumbers={ multiples }
        onNumberClickFn={ onNumberSelectedFn }
      />
    </>
  );
};

MultiplicationGrid.propTypes = {
  highestMultiplierNumber: PropTypes.number
};

MultiplicationGrid.defaultProps = {
  highestMultiplierNumber: 0
};

export default MultiplicationGrid;