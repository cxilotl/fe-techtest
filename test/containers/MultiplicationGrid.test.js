import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, fireEvent, waitForElement } from '@testing-library/react';
import MultiplicationGrid from '../../src/containers/MultiplicationGrid';

describe('MultiplicationGrid', () => {
  const maxNumber = 25;

  it('snapshot', () => {
    const component = renderer.create(<MultiplicationGrid highestMultiplierNumber={ maxNumber } />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Given a list of numbers for which we want to know the multiples of a selected number within the list, it', () => {
    let highestNum = 25;

    afterEach(cleanup);

    it('Should render an unselected list of buttons', () => {
      const { container } = render(
        <MultiplicationGrid highestMultiplierNumber={ maxNumber } />
      );
      const buttons = container.querySelectorAll('button');
      expect(buttons.length).toEqual(highestNum - 1);
    });

    it('Should render a non highlighted list of buttons, when no specific number was chosen', () => {
      const { container } = render(
        <MultiplicationGrid highestMultiplierNumber={ maxNumber } />
      );
      const highlightedButtons = container.querySelectorAll('button.isHighlighted');
      expect(highlightedButtons.length).toEqual(0);
    });

    it('Should render a highlighted list of buttons when a specific number is selected', async () => {
      const { container, getByText } = render(
        <MultiplicationGrid highestMultiplierNumber={ maxNumber }  />
      );

      // Selecting the multiplier button and clicking on it
      const clickedButton = getByText('5');
      fireEvent.click(clickedButton);

      // Checking the multiplier has been selected in the app
      const selectedButtons = await waitForElement(() => container.querySelectorAll('.isSelected'), { container });
      expect(selectedButtons.length).toEqual(1);
      expect(parseInt(selectedButtons[0].value, 10)).toEqual(5);

      // Checking that the multiples have been highlighted
      const multiplesHighlightedButtons = await waitForElement(() => container.querySelectorAll('.isHighlighted'), { container });
      expect(multiplesHighlightedButtons.length).toEqual(3);
      expect(parseInt(multiplesHighlightedButtons[0].value, 10)).toEqual(10);
      expect(parseInt(multiplesHighlightedButtons[1].value, 10)).toEqual(15);
      expect(parseInt(multiplesHighlightedButtons[2].value, 10)).toEqual(20);
    });
  });
});