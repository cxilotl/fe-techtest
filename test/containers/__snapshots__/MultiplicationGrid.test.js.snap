// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiplicationGrid snapshot 1`] = `
<ol
  className="layout"
>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-1"
      onClick={[Function]}
      type="button"
      value={1}
    >
      1
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-2"
      onClick={[Function]}
      type="button"
      value={2}
    >
      2
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-3"
      onClick={[Function]}
      type="button"
      value={3}
    >
      3
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-4"
      onClick={[Function]}
      type="button"
      value={4}
    >
      4
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-5"
      onClick={[Function]}
      type="button"
      value={5}
    >
      5
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-6"
      onClick={[Function]}
      type="button"
      value={6}
    >
      6
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-7"
      onClick={[Function]}
      type="button"
      value={7}
    >
      7
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-8"
      onClick={[Function]}
      type="button"
      value={8}
    >
      8
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-9"
      onClick={[Function]}
      type="button"
      value={9}
    >
      9
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-10"
      onClick={[Function]}
      type="button"
      value={10}
    >
      10
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-11"
      onClick={[Function]}
      type="button"
      value={11}
    >
      11
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-12"
      onClick={[Function]}
      type="button"
      value={12}
    >
      12
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-13"
      onClick={[Function]}
      type="button"
      value={13}
    >
      13
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-14"
      onClick={[Function]}
      type="button"
      value={14}
    >
      14
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-15"
      onClick={[Function]}
      type="button"
      value={15}
    >
      15
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-16"
      onClick={[Function]}
      type="button"
      value={16}
    >
      16
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-17"
      onClick={[Function]}
      type="button"
      value={17}
    >
      17
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-18"
      onClick={[Function]}
      type="button"
      value={18}
    >
      18
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-19"
      onClick={[Function]}
      type="button"
      value={19}
    >
      19
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-20"
      onClick={[Function]}
      type="button"
      value={20}
    >
      20
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-21"
      onClick={[Function]}
      type="button"
      value={21}
    >
      21
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-22"
      onClick={[Function]}
      type="button"
      value={22}
    >
      22
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-23"
      onClick={[Function]}
      type="button"
      value={23}
    >
      23
    </button>
  </li>
  <li
    className="layout"
  >
    <button
      className="layout"
      name="num-button-24"
      onClick={[Function]}
      type="button"
      value={24}
    >
      24
    </button>
  </li>
</ol>
`;
